realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java

```diff
@@ -0,0 +1,423 @@
+        assertEquals(1, child.getObjectParents().size());
+        assertTrue(child.getObjectParents().contains(parent));
+    }
+
+    // Fields annotated with @LinkingObjects should not be affected by JSON updates
+    @Test
+    public void jsonUpdate_list() {
+        realm.beginTransaction();
+        AllJavaTypes child = realm.createObject(AllJavaTypes.class, 1);
+        AllJavaTypes parent = realm.createObject(AllJavaTypes.class, 2);
+        parent.getFieldList().add(child);
+        realm.commitTransaction();
+
+        realm.beginTransaction();
+        try {
+            realm.createOrUpdateAllFromJson(AllJavaTypes.class, "[{ \"fieldId\" : 1, \"listParents\" : null }]");
+            fail("Expected attempt to load the @LinkingObjects 'listParents' field to fail");
+        } catch (RealmException ignore) {
+        }
+        realm.commitTransaction();
+
+        RealmResults<AllJavaTypes> parents = child.getListParents();
+        assertNotNull(parents);
+        assertEquals(1, child.getListParents().size());
+        assertTrue(child.getListParents().contains(parent));
+    }
+
+    // Distinct works for backlinks
+    @Test
+    public void multipleReferencesWithDistinct() {
+        realm.beginTransaction();
+        AllJavaTypes child = realm.createObject(AllJavaTypes.class, 1);
+        AllJavaTypes parent = realm.createObject(AllJavaTypes.class, 2);
+        parent.getFieldList().add(child);
+        parent.getFieldList().add(child);
+        realm.commitTransaction();
+
+        assertEquals(2, child.getListParents().size());
+
+        RealmResults<AllJavaTypes> distinctParents = child.getListParents().where().distinct("fieldId");
+        assertEquals(1, distinctParents.size());
+        assertTrue(child.getListParents().contains(parent));
+    }
+
+    // Query on a field descriptor starting with a backlink
+    @Test
+    @Ignore
+    public void queryStartingWithBacklink() {
+        realm.beginTransaction();
+        AllJavaTypes gen1 = realm.createObject(AllJavaTypes.class, 10);
+
+        AllJavaTypes gen2A = realm.createObject(AllJavaTypes.class, 1);
+        gen2A.setFieldObject(gen1);
+
+        AllJavaTypes gen2B = realm.createObject(AllJavaTypes.class, 2);
+        gen2B.setFieldObject(gen1);
+
+        AllJavaTypes gen3 = realm.createObject(AllJavaTypes.class, 3);
+        RealmList<AllJavaTypes> parents = gen3.getFieldList();
+        parents.add(gen2A);
+        parents.add(gen2B);
+
+        realm.commitTransaction();
+
+        RealmResults<AllJavaTypes> result = realm.where(AllJavaTypes.class)
+            .greaterThan("objectParents.fieldId", 1)
+            .findAll();
+        assertEquals(1, result.size());
+        assertTrue(result.contains(gen2B));
+    }
+
+    // Query on a field descriptor that ends with a backlink
+    @Test
+    @Ignore
+    public void queryEndingWithBacklink() {
+        realm.beginTransaction();
+        AllJavaTypes gen1 = realm.createObject(AllJavaTypes.class, 10);
```

Comment:
`gen1/gen2A/gen2B/gen3` is fairly confusing. Either rename or add comment about what it means.